lib_xua/src/core/endpoint0/xua_endpoint0.c

```diff
@@ -951,13 +1000,103 @@ void XUA_Endpoint0_loop(XUD_Result_t result, USB_SetupPacket_t sp, chanend c_ep0
     } /* if(result == XUD_RES_OKAY) */
 
     {
+#if _XUA_ENABLE_BOS_DESC
+        if(result == XUD_RES_ERR)
+        {
+            unsigned bmRequestType = (sp.bmRequestType.Direction<<7) | (sp.bmRequestType.Type<<5) | (sp.bmRequestType.Recipient);
+            if(bmRequestType == USB_BMREQ_D2H_VENDOR_DEV)
+            {
+                if((sp.bRequest == REQUEST_GET_MS_DESCRIPTOR) &&
+                    sp.wIndex == MS_OS_20_DESCRIPTOR_INDEX)
+                {
+                    int num_interfaces;
+                    if(DFU_mode_active) {
+                        num_interfaces = DFUcfgDesc.Config.bNumInterfaces;
+                    }
+                    else {
+                        num_interfaces = cfgDesc_Audio2.Config.bNumInterfaces;
+                    }
+
+                    if(num_interfaces == 1) {
+                        result = XUD_DoGetRequest(ep0_out, ep0_in, (unsigned char*)&desc_ms_os_20_simple, sizeof(MSOS_desc_simple_t), sp.wLength);
+                    }
+                    else {
+                        result = XUD_DoGetRequest(ep0_out, ep0_in, (unsigned char*)&desc_ms_os_20_composite, sizeof(MSOS_desc_composite_t), sp.wLength);
+                    }
+                }
+            }
+
+        }
+#endif
+
+#if XUA_DFU_EN
+        if(result == XUD_RES_ERR)
+        {
+            // From Theycon DFU driver v2.66.0 onwards, the XMOS_DFU_REVERTFACTORY request comes as a H2D vendor request and not as a class request addressed to the DFU interface
+            unsigned bmRequestType = (sp.bmRequestType.Direction<<7) | (sp.bmRequestType.Type<<5) | (sp.bmRequestType.Recipient);
```

Comment:
Given this is an interface request, shouldn't we strictly check the interface number in the request? (sp.wIndex & 0xff)

Reply:
oops, sorrry I missed that. You're right, interface number should be check. Normally, the revertfactory request comes when in DFU mode but we can't depend on that so set the dfu interface number based on the DFU_mode_active flag.